R/calcEmiTarget.R

```diff
@@ -49,18 +67,71 @@ calcEmiTarget <- function(sources, subtype, scenario) {
   # ensure that all magclass objects in the list have matching years so they can be bound together
-
   listYears <- lapply(listGhgFactors, getItems, dim = "year") %>% unlist() %>% unique() %>% sort()
   ghgFactor <- purrr::map(listGhgFactors,
                           ~ add_columns(.x, listYears[!listYears %in% getItems(.x, dim = "year")], 2))
-
   ghgFactor <- mbind(ghgFactor)
   ghgFactor <- ghgFactor[, sort(getYears(ghgFactor)), ]
 
-  # GHG factors weighted by GHG in 2015
-  if (subtype == "Ghgfactor") {
 
+  # 2. Extrapolate NDC targets from 2030 to 2035 for countries which do not have 2035 NDC targets ----
+
+  # add NDC versions "Emissions_NDC " and "Emissions_2025_uncond_extrapol"
+  # that take the 2035 targets from the 2025 version, but extrapolate 2030 targets to 2035 for countries which do not have 2035 targets yet
+  # create new NDC version 2025 with extrapolated targets
+
+  # if no 2035 targets available, add 2035 to target factor object
+  if (! "y2035" %in% getYears(ghgFactor)) {
+    ghgFactor <- mbind(ghgFactor, new.magpie(getRegions(ghgFactor),"y2035",getNames(ghgFactor), fill = NA))
+  }
+  ghgFactorExtrapolated <- ghgFactor[,,c("2025_cond","2025_uncond")]
+  getItems(ghgFactorExtrapolated, dim=3.1) <- paste0(getItems(ghgFactorExtrapolated, dim=3.1),"_extrapol")
+  # loop over countries and NDC scenarios to extrapolate 2030 targets to 2035 targets for those countries who do not have 2035 targets
+  for (country in getRegions(ghgFactorExtrapolated ) ) {
+    for (NDCscenario in getItems(ghgFactorExtrapolated, dim=3) ) {
+      # do extrapolation only if country has 2030 target but no 2035 target
+      if (c("y2035") %in% getYears(ghgFactorExtrapolated)) {
+        Do2035Extrapolation <- ( as.vector(is.na(ghgFactorExtrapolated[country, "y2035", NDCscenario])) &
+                                 as.vector(!is.na(ghgFactorExtrapolated[country, "y2030", NDCscenario])))
+      } else {
+        Do2035Extrapolation <- as.vector(!is.na(ghgFactorExtrapolated[country, "y2030", NDCscenario]))
+      }
+
+      # if country has 2030 target but no 2035 target
+      if (Do2035Extrapolation) {
+        # if no 2035 target available, linearly extrapolate target from 2030 to 2035
+        # Explanation of the calculation:
+        # Note that the ghgFactor gives the remaining relative emissions relative to 2015.
+        # Hence, 1 - ghgFactor gives the relative emissions reductions relative to 2015.
+        # Dividing by 15 years gives annual average emissions reductions.
+        # Multiplying by 20 years applies these emissions reductions over the whole 20-year period from 2015 to 2035.
+        # These relative emissions reduction need to be converted to a target factor by subtracting them from one.
+        ghgFactorExtrapolated[country, "y2035", NDCscenario] <- 1 - ( (1 - ghgFactorExtrapolated[country, "y2030", NDCscenario]) / 15 * 20)
+
+        print(paste0(country, " has no 2035 NDC emissions target."))
```

Comment:
I know these are useful for development, but clutter up the input data log. Consider removing the print statement once you are certain your code works as expected, or convert them into messages.